// ionic/dist/commands/doctor/ignore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk = require("chalk");
const command_1 = require("@ionic/cli-utils/lib/command");
const errors_1 = require("@ionic/cli-utils/lib/errors");
let DoctorIgnoreCommand = class DoctorIgnoreCommand extends command_1.Command {
    preRun(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { Ailments } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/doctor/ailments'); });
            const { contains, validate } = yield Promise.resolve().then(function () { return require('@ionic/cli-utils/lib/validators'); });
            const ailmentIds = Ailments.ALL.map(Ailment => new Ailment().id);
            if (options['list']) {
                this.env.log.msg(ailmentIds.map(id => chalk.green(id)).join('\n'));
                throw new errors_1.FatalException('', 0);
            }
            if (!inputs[0]) {
                inputs[0] = yield this.env.prompt({
                    type: 'list',
                    name: 'id',
                    message: 'Which issue would you like to ignore?',
                    choices: ailmentIds,
                });
            }
            validate(inputs[0], 'id', [contains(ailmentIds, {})]);
        });
    }
    run(inputs, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const [id] = inputs;
            const config = yield this.env.config.load();
            config.state.doctor.ignored.push(id);
            this.env.log.ok(`Ignored issue ${chalk.green(id)}`);
        });
    }
};
DoctorIgnoreCommand = tslib_1.__decorate([
    command_1.CommandMetadata({
        name: 'ignore',
        type: 'project',
        description: 'Ignore a particular issue',
        visible: false,
        exampleCommands: [
            '',
            '--list',
        ],
        inputs: [
            {
                name: 'id',
                description: 'The issue identifier',
            },
        ],
        options: [
            {
                name: 'list',
                description: 'List issue identifiers',
                type: Boolean,
                aliases: ['l'],
            },
        ],
    })
], DoctorIgnoreCommand);
exports.DoctorIgnoreCommand = DoctorIgnoreCommand;
